mkvcore/blockreader.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2019 The ebml-go authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mkvcore
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+	"sync"
+
+	"github.com/at-wat/ebml-go"
+)
+
+type ReadableBlock struct {
+	Header      interface{}
+	trackNumber uint64
+	f           chan *frame
+	wg          *sync.WaitGroup
+}
+
+func (r *ReadableBlock) Read() (b []byte, keyframe bool, timestamp int64, err error) {
+	frame := <-r.f
+	return frame.b, frame.keyframe, frame.timestamp, nil
+}
+
+func (r *ReadableBlock) Close() error {
+	r.wg.Done()
+
+	return nil
+}
+
+// NewSimpleBlockReader creates BlockReadCloser for each track specified as tracks argument.
+func NewSimpleBlockReader(r0 io.Reader, options BlockReaderOptions) ([]ReadableBlock, error) {
```

Comment:
```suggestion
func NewSimpleBlockReader(r0 io.Reader, options ...BlockReaderOption) ([]BlockReadCloser, error) {
```
Please return readers as `BlockReadCloser interface` instead of `ReadableBlock struct` and receive options using the functional options pattern.